test/e2e/dr/quorum-restore.go

```diff
@@ -0,0 +1,330 @@
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+)
+
+const (
+	machineAnnotationName   = "machine.openshift.io/machine"
+	localEtcdSignerYaml     = "/tmp/kube-etcd-cert-signer.yaml"
+	expectedNumberOfMasters = 3
+)
+
+var _ = g.Describe("[Feature:DisasterRecovery][Disruptive]", func() {
+	f := e2e.NewDefaultFramework("disaster-recovery")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("disaster-recovery")
+
+	g.It("[dr-quorum-restore] Cluster should restore itself after quorum loss", func() {
+		config, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dynamicClient := dynamic.NewForConfigOrDie(config)
+		ms := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "machine.openshift.io",
+			Version:  "v1beta1",
+			Resource: "machines",
+		}).Namespace("openshift-machine-api")
+		mcps := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "machineconfiguration.openshift.io",
+			Version:  "v1",
+			Resource: "machineconfigpools",
+		})
+		coc := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "config.openshift.io",
+			Version:  "v1",
+			Resource: "clusteroperators",
+		})
+
+		bastionHost := setupSSHBastion(oc)
+		proxy := fmt.Sprintf(proxyTemplate, sshOpts, bastionHost)
+		defer removeSSHBastion(oc)
+
+		scaleEtcdQuorum(oc, 0)
+
+		e2e.Logf("Finding two masters to remove")
+		mapiPods, err := oc.AdminKubeClient().CoreV1().Pods("openshift-machine-api").List(metav1.ListOptions{
+			LabelSelector: "k8s-app=controller",
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(mapiPods.Items).NotTo(o.BeEmpty())
+
+		survivingNodeName := mapiPods.Items[0].Spec.NodeName
+		mastersNodes := getAllMasters(oc)
+		o.Expect(mastersNodes).NotTo(o.BeEmpty())
+
+		survivingMachineName := getMachineNameByNodeName(oc, survivingNodeName)
+		survivingMachine, err := ms.Get(survivingMachineName, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// Set etcd connection string before destroying masters, as ssh bastion may be unavailable
+		etcdConnectionString := constructEtcdConnectionString([]string{survivingNodeName}, proxy)
+
+		e2e.Logf("Destroy 2 masters")
+		masterMachines := make([]string, len(mastersNodes))
+		for i, node := range mastersNodes {
+			masterMachine := getMachineNameByNodeName(oc, node)
+			masterMachines[i] = masterMachine
+
+			if node == survivingNodeName {
+				continue
+			}
+
+			e2e.Logf("Destroying %s", masterMachine)
+			err = ms.Delete(masterMachine, &metav1.DeleteOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+		e2e.Logf("masterMachines: %v", masterMachines)
+
+		e2e.Logf("Confirm meltdown")
+		time.Sleep(30 * time.Second)
+		err = wait.Poll(30*time.Second, 30*time.Minute, func() (done bool, err error) {
+			_, err = oc.AdminKubeClient().CoreV1().Nodes().List(metav1.ListOptions{})
+			return err != nil, nil
+		})
+
+		e2e.Logf("Restore single node etcd")
```

Comment:
Seems like from here down could have been a script.  I won't hold the pull on it, but all these steps are now going to have to be documented without a clear example or clear path towards something productized, right?

Reply:
Right, that's [documented](https://docs.openshift.com/container-platform/4.1/disaster_recovery/scenario-1-infra-recovery.html) ("Set the INITIAL_CLUSTER variable" and "Run the etcd-snapshot-restore.sh script") and its one-liner over ssh bastion.